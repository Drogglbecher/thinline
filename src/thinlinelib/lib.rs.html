<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Source to the Rust file `src/lib.rs`."><meta name="keywords" content="rust, rustlang, rust-lang"><title>lib.rs.html -- source</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css"><link rel="stylesheet" type="text/css" href="../../light.css" id="themeStyle"><script src="../../storage.js"></script></head><body class="rustdoc source"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!"><img src="../../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices"></div></div><script src="../../theme.js"></script><nav class="sub"><form class="search-form js-only"><div class="search-container"><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><a id="settings-menu" href="../../settings.html"><img src="../../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><pre class="line-numbers"><span id="1">  1</span>
<span id="2">  2</span>
<span id="3">  3</span>
<span id="4">  4</span>
<span id="5">  5</span>
<span id="6">  6</span>
<span id="7">  7</span>
<span id="8">  8</span>
<span id="9">  9</span>
<span id="10"> 10</span>
<span id="11"> 11</span>
<span id="12"> 12</span>
<span id="13"> 13</span>
<span id="14"> 14</span>
<span id="15"> 15</span>
<span id="16"> 16</span>
<span id="17"> 17</span>
<span id="18"> 18</span>
<span id="19"> 19</span>
<span id="20"> 20</span>
<span id="21"> 21</span>
<span id="22"> 22</span>
<span id="23"> 23</span>
<span id="24"> 24</span>
<span id="25"> 25</span>
<span id="26"> 26</span>
<span id="27"> 27</span>
<span id="28"> 28</span>
<span id="29"> 29</span>
<span id="30"> 30</span>
<span id="31"> 31</span>
<span id="32"> 32</span>
<span id="33"> 33</span>
<span id="34"> 34</span>
<span id="35"> 35</span>
<span id="36"> 36</span>
<span id="37"> 37</span>
<span id="38"> 38</span>
<span id="39"> 39</span>
<span id="40"> 40</span>
<span id="41"> 41</span>
<span id="42"> 42</span>
<span id="43"> 43</span>
<span id="44"> 44</span>
<span id="45"> 45</span>
<span id="46"> 46</span>
<span id="47"> 47</span>
<span id="48"> 48</span>
<span id="49"> 49</span>
<span id="50"> 50</span>
<span id="51"> 51</span>
<span id="52"> 52</span>
<span id="53"> 53</span>
<span id="54"> 54</span>
<span id="55"> 55</span>
<span id="56"> 56</span>
<span id="57"> 57</span>
<span id="58"> 58</span>
<span id="59"> 59</span>
<span id="60"> 60</span>
<span id="61"> 61</span>
<span id="62"> 62</span>
<span id="63"> 63</span>
<span id="64"> 64</span>
<span id="65"> 65</span>
<span id="66"> 66</span>
<span id="67"> 67</span>
<span id="68"> 68</span>
<span id="69"> 69</span>
<span id="70"> 70</span>
<span id="71"> 71</span>
<span id="72"> 72</span>
<span id="73"> 73</span>
<span id="74"> 74</span>
<span id="75"> 75</span>
<span id="76"> 76</span>
<span id="77"> 77</span>
<span id="78"> 78</span>
<span id="79"> 79</span>
<span id="80"> 80</span>
<span id="81"> 81</span>
<span id="82"> 82</span>
<span id="83"> 83</span>
<span id="84"> 84</span>
<span id="85"> 85</span>
<span id="86"> 86</span>
<span id="87"> 87</span>
<span id="88"> 88</span>
<span id="89"> 89</span>
<span id="90"> 90</span>
<span id="91"> 91</span>
<span id="92"> 92</span>
<span id="93"> 93</span>
<span id="94"> 94</span>
<span id="95"> 95</span>
<span id="96"> 96</span>
<span id="97"> 97</span>
<span id="98"> 98</span>
<span id="99"> 99</span>
<span id="100">100</span>
<span id="101">101</span>
<span id="102">102</span>
<span id="103">103</span>
<span id="104">104</span>
<span id="105">105</span>
<span id="106">106</span>
<span id="107">107</span>
<span id="108">108</span>
<span id="109">109</span>
<span id="110">110</span>
<span id="111">111</span>
<span id="112">112</span>
<span id="113">113</span>
<span id="114">114</span>
<span id="115">115</span>
<span id="116">116</span>
<span id="117">117</span>
<span id="118">118</span>
<span id="119">119</span>
<span id="120">120</span>
<span id="121">121</span>
<span id="122">122</span>
<span id="123">123</span>
<span id="124">124</span>
<span id="125">125</span>
<span id="126">126</span>
<span id="127">127</span>
<span id="128">128</span>
<span id="129">129</span>
<span id="130">130</span>
<span id="131">131</span>
<span id="132">132</span>
<span id="133">133</span>
<span id="134">134</span>
<span id="135">135</span>
<span id="136">136</span>
<span id="137">137</span>
<span id="138">138</span>
<span id="139">139</span>
<span id="140">140</span>
<span id="141">141</span>
<span id="142">142</span>
<span id="143">143</span>
<span id="144">144</span>
<span id="145">145</span>
<span id="146">146</span>
<span id="147">147</span>
<span id="148">148</span>
<span id="149">149</span>
<span id="150">150</span>
<span id="151">151</span>
<span id="152">152</span>
<span id="153">153</span>
<span id="154">154</span>
<span id="155">155</span>
<span id="156">156</span>
<span id="157">157</span>
<span id="158">158</span>
<span id="159">159</span>
<span id="160">160</span>
<span id="161">161</span>
<span id="162">162</span>
<span id="163">163</span>
<span id="164">164</span>
<span id="165">165</span>
<span id="166">166</span>
<span id="167">167</span>
<span id="168">168</span>
<span id="169">169</span>
<span id="170">170</span>
<span id="171">171</span>
<span id="172">172</span>
<span id="173">173</span>
<span id="174">174</span>
<span id="175">175</span>
<span id="176">176</span>
<span id="177">177</span>
<span id="178">178</span>
<span id="179">179</span>
<span id="180">180</span>
<span id="181">181</span>
<span id="182">182</span>
<span id="183">183</span>
<span id="184">184</span>
<span id="185">185</span>
<span id="186">186</span>
<span id="187">187</span>
<span id="188">188</span>
<span id="189">189</span>
<span id="190">190</span>
<span id="191">191</span>
<span id="192">192</span>
<span id="193">193</span>
</pre><pre class="rust ">
<span class="attribute">#![<span class="ident">cfg_attr</span>(<span class="ident">feature</span> <span class="op">=</span> <span class="string">&quot;clippy&quot;</span>, <span class="ident">feature</span>(<span class="ident">plugin</span>))]</span>
<span class="attribute">#![<span class="ident">cfg_attr</span>(<span class="ident">feature</span> <span class="op">=</span> <span class="string">&quot;clippy&quot;</span>, <span class="ident">plugin</span>(<span class="ident">clippy</span>))]</span>

<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">clang</span>;
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">directories</span>;
<span class="attribute">#[<span class="ident">macro_use</span>]</span>
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">failure</span>;
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">glob</span>;
<span class="attribute">#[<span class="ident">macro_use</span>]</span>
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">lazy_static</span>;
<span class="attribute">#[<span class="ident">macro_use</span>]</span>
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">log</span>;
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">python_parser</span>;
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">regex</span>;
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">run_script</span>;
<span class="attribute">#[<span class="ident">macro_use</span>]</span>
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">serde_derive</span>;
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">serde</span>;
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">slog_envlogger</span>;
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">snapshot</span>;
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">walkdir</span>;
<span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">yaml_rust</span>;

<span class="kw">pub</span> <span class="kw">mod</span> <span class="ident">analysis</span>;
<span class="kw">pub</span> <span class="kw">mod</span> <span class="ident">entity</span>;
<span class="kw">pub</span> <span class="kw">mod</span> <span class="ident">language_type</span>;
<span class="kw">pub</span> <span class="kw">mod</span> <span class="ident">project_parameters</span>;
<span class="kw">pub</span> <span class="kw">mod</span> <span class="ident">stubs</span>;
<span class="kw">pub</span> <span class="kw">mod</span> <span class="ident">synthesis</span>;
<span class="kw">pub</span> <span class="kw">mod</span> <span class="ident">value_parser</span>;

<span class="kw">use</span> <span class="ident">analysis</span>::{<span class="ident">Analysis</span>, <span class="ident">ProjectFile</span>};
<span class="kw">use</span> <span class="ident">directories</span>::<span class="ident">BaseDirs</span>;
<span class="kw">use</span> <span class="ident">failure</span>::{<span class="ident">err_msg</span>, <span class="ident">Fallible</span>};
<span class="kw">use</span> <span class="ident">language_type</span>::<span class="ident">LanguageType</span>;
<span class="kw">use</span> <span class="ident">project_parameters</span>::<span class="ident">ProjectParameters</span>;
<span class="kw">use</span> <span class="ident">std</span>::<span class="ident">path</span>::<span class="ident">PathBuf</span>;
<span class="kw">use</span> <span class="ident">synthesis</span>::<span class="kw-2">*</span>;

<span class="comment">////////////////////////////////////////////////////////////////////////////////</span>

<span class="kw">static</span> <span class="ident">DEFAULT_ENV_YML</span>: <span class="kw-2">&amp;</span><span class="ident">str</span> <span class="op">=</span> <span class="string">&quot;./stubs/environment/env_stubs.yml&quot;</span>;

<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Default</span>, <span class="ident">Debug</span>)]</span>
<span class="doccomment">/// Global structure representing the `Thinline` lib.</span>
<span class="kw">pub</span> <span class="kw">struct</span> <span class="ident">Thinline</span><span class="op">&lt;</span><span class="ident">T</span><span class="op">&gt;</span>
<span class="kw">where</span>
    <span class="ident">T</span>: <span class="ident">LanguageType</span>,
{
    <span class="doccomment">/// The target project directory.</span>
    <span class="kw">pub</span> <span class="ident">project_dir</span>: <span class="ident">PathBuf</span>,

    <span class="doccomment">/// The parsed project parameters.</span>
    <span class="kw">pub</span> <span class="ident">project_parameters</span>: <span class="ident">ProjectParameters</span>,

    <span class="doccomment">/// The structure holding the analysis_c data.</span>
    <span class="ident">analysis</span>: <span class="ident">Analysis</span><span class="op">&lt;</span><span class="ident">T</span><span class="op">&gt;</span>,

    <span class="doccomment">/// The structure holding the synthesized testdata.</span>
    <span class="ident">synthesis</span>: <span class="ident">Synthesis</span><span class="op">&lt;</span><span class="ident">T</span><span class="op">&gt;</span>,
}

<span class="kw">impl</span><span class="op">&lt;</span><span class="ident">T</span><span class="op">&gt;</span> <span class="ident">Thinline</span><span class="op">&lt;</span><span class="ident">T</span><span class="op">&gt;</span>
<span class="kw">where</span>
    <span class="ident">T</span>: <span class="ident">LanguageType</span>,
{
    <span class="doccomment">/// Creates an instance of the lib containing `Thinline`s functionality.</span>
    <span class="kw">pub</span> <span class="kw">fn</span> <span class="ident">new</span><span class="op">&lt;</span><span class="ident">P</span>: <span class="ident">Into</span><span class="op">&lt;</span><span class="ident">PathBuf</span><span class="op">&gt;&gt;</span>(<span class="ident">project_dir</span>: <span class="ident">P</span>) <span class="op">-&gt;</span> <span class="self">Self</span> {
        <span class="self">Self</span> {
            <span class="ident">project_dir</span>: <span class="ident">project_dir</span>.<span class="ident">into</span>(),
            <span class="ident">project_parameters</span>: <span class="ident">ProjectParameters</span>::<span class="ident">new</span>(),
            <span class="ident">analysis</span>: <span class="ident">Analysis</span>::<span class="ident">new</span>(),
            <span class="ident">synthesis</span>: <span class="ident">Synthesis</span>::<span class="ident">new</span>(),
        }
    }

    <span class="doccomment">/// Starts the analysis of the target project.</span>
    <span class="kw">pub</span> <span class="kw">fn</span> <span class="ident">analyze</span>(<span class="kw-2">&amp;</span><span class="kw-2">mut</span> <span class="self">self</span>, <span class="ident">thinline_cfg</span>: <span class="kw-2">&amp;</span><span class="ident">str</span>, <span class="ident">build</span>: <span class="ident">bool</span>) <span class="op">-&gt;</span> <span class="ident">Fallible</span><span class="op">&lt;</span>()<span class="op">&gt;</span> {
        <span class="comment">// Parses the project config</span>
        <span class="self">self</span>.<span class="ident">parse_project_config</span>(<span class="ident">thinline_cfg</span>)<span class="question-mark">?</span>;

        <span class="comment">// Analyze the project at the given source directory</span>
        <span class="self">self</span>.<span class="ident">analyze_project</span>()<span class="question-mark">?</span>;

        <span class="comment">// Builds target project when build flag is set</span>
        <span class="kw">if</span> <span class="ident">build</span> {
            <span class="self">self</span>.<span class="ident">project_parameters</span>.<span class="ident">build_script</span>.<span class="ident">run</span>(<span class="kw-2">&amp;</span><span class="self">self</span>.<span class="ident">project_dir</span>)<span class="question-mark">?</span>;
        }

        <span class="prelude-val">Ok</span>(())
    }

    <span class="doccomment">/// Traverses upwared through the `seed_dir`s parent directories until</span>
    <span class="doccomment">/// `thinline.yml` configuration exists in dir or home directory is reached.</span>
    <span class="kw">fn</span> <span class="ident">find_thinline_config_dir</span><span class="op">&lt;</span><span class="lifetime">&#39;a</span><span class="op">&gt;</span>(<span class="ident">seed_dir</span>: <span class="kw-2">&amp;</span><span class="lifetime">&#39;a</span> <span class="ident">PathBuf</span>, <span class="ident">config_name</span>: <span class="kw-2">&amp;</span><span class="ident">str</span>) <span class="op">-&gt;</span> <span class="prelude-ty">Option</span><span class="op">&lt;</span><span class="ident">PathBuf</span><span class="op">&gt;</span> {
        <span class="kw">if</span> <span class="kw">let</span> <span class="prelude-val">Some</span>(<span class="ident">base_dirs</span>) <span class="op">=</span> <span class="ident">BaseDirs</span>::<span class="ident">new</span>() {
            <span class="kw">let</span> <span class="ident">home_dir</span> <span class="op">=</span> <span class="ident">base_dirs</span>.<span class="ident">home_dir</span>();
            <span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">project_config</span> <span class="op">=</span> <span class="ident">seed_dir</span>.<span class="ident">clone</span>();

            <span class="kw">loop</span> {
                <span class="kw">if</span> <span class="ident">project_config</span> <span class="op">==</span> <span class="ident">home_dir</span> <span class="op">||</span> <span class="ident">project_config</span>.<span class="ident">join</span>(<span class="ident">config_name</span>).<span class="ident">is_file</span>() {
                    <span class="macro">trace</span><span class="macro">!</span>(<span class="string">&quot;Config dir: {:?}&quot;</span>, <span class="ident">project_config</span>);
                    <span class="kw">break</span>;
                }

                <span class="kw">if</span> <span class="kw">let</span> <span class="prelude-val">Some</span>(<span class="ident">parent</span>) <span class="op">=</span> <span class="ident">project_config</span>.<span class="ident">clone</span>().<span class="ident">as_path</span>().<span class="ident">parent</span>() {
                    <span class="ident">project_config</span> <span class="op">=</span> <span class="ident">parent</span>.<span class="ident">to_path_buf</span>();
                } <span class="kw">else</span> {
                    <span class="kw">return</span> <span class="prelude-val">None</span>;
                }
            }

            <span class="kw">return</span> <span class="prelude-val">Some</span>(<span class="ident">project_config</span>.<span class="ident">join</span>(<span class="ident">config_name</span>));
        }

        <span class="prelude-val">None</span>
    }

    <span class="doccomment">/// Parses configuration from the given config yaml.</span>
    <span class="kw">fn</span> <span class="ident">parse_project_config</span>(<span class="kw-2">&amp;</span><span class="kw-2">mut</span> <span class="self">self</span>, <span class="ident">config_name</span>: <span class="kw-2">&amp;</span><span class="ident">str</span>) <span class="op">-&gt;</span> <span class="ident">Fallible</span><span class="op">&lt;</span>()<span class="op">&gt;</span> {
        <span class="kw">if</span> <span class="kw">let</span> <span class="prelude-val">Some</span>(<span class="ident">project_config</span>) <span class="op">=</span> <span class="self">Self</span>::<span class="ident">find_thinline_config_dir</span>(<span class="kw-2">&amp;</span><span class="self">self</span>.<span class="ident">project_dir</span>, <span class="ident">config_name</span>)
        {
            <span class="kw">if</span> <span class="op">!</span><span class="ident">project_config</span>.<span class="ident">exists</span>() <span class="op">||</span> <span class="op">!</span><span class="ident">project_config</span>.<span class="ident">is_file</span>() {
                <span class="kw">return</span> <span class="prelude-val">Err</span>(<span class="macro">format_err</span><span class="macro">!</span>(
                    <span class="string">&quot;The given project config file &#39;{}&#39; does not exist or is a directory.&quot;</span>,
                    <span class="ident">project_config</span>
                        .<span class="ident">to_str</span>()
                        .<span class="ident">ok_or_else</span>(<span class="op">||</span> <span class="ident">err_msg</span>(<span class="string">&quot;Unable to stringify project config file.&quot;</span>))<span class="question-mark">?</span>
                ));
            }

            <span class="self">self</span>.<span class="ident">project_parameters</span> <span class="op">=</span> <span class="ident">ProjectParameters</span>::<span class="ident">parse</span>(
                <span class="ident">project_config</span>
                    .<span class="ident">to_str</span>()
                    .<span class="ident">ok_or_else</span>(<span class="op">||</span> <span class="ident">err_msg</span>(<span class="string">&quot;Unable to stringify project config file.&quot;</span>))<span class="question-mark">?</span>,
            )<span class="question-mark">?</span>;

            <span class="macro">debug</span><span class="macro">!</span>(<span class="string">&quot;{:#?}&quot;</span>, <span class="self">self</span>.<span class="ident">project_parameters</span>);
        } <span class="kw">else</span> {
            <span class="kw">return</span> <span class="prelude-val">Err</span>(<span class="macro">format_err</span><span class="macro">!</span>(<span class="string">&quot;Unable to get project dir parent for&quot;</span>));
        }

        <span class="prelude-val">Ok</span>(())
    }

    <span class="doccomment">/// Analyzes the project which should be tested.</span>
    <span class="kw">fn</span> <span class="ident">analyze_project</span>(<span class="kw-2">&amp;</span><span class="self">self</span>) <span class="op">-&gt;</span> <span class="ident">Fallible</span><span class="op">&lt;</span>()<span class="op">&gt;</span> {
        <span class="kw">if</span> <span class="kw">let</span> <span class="prelude-val">Some</span>(<span class="ident">project_path_s</span>) <span class="op">=</span> <span class="self">self</span>.<span class="ident">project_dir</span>.<span class="ident">to_str</span>() {
            <span class="macro">info</span><span class="macro">!</span>(<span class="string">&quot;Starting project analysis at &#39;{}&#39;&quot;</span>, <span class="ident">project_path_s</span>);
        }

        <span class="kw">if</span> <span class="self">self</span>.<span class="ident">project_dir</span>.<span class="ident">is_dir</span>() {
            <span class="comment">// Project path is a directory, thus it is neccessay to traverse to the project</span>
            <span class="comment">// and collect all the sources.</span>
            <span class="self">self</span>.<span class="ident">analysis</span>
                .<span class="ident">collect_sources</span>(<span class="kw-2">&amp;</span><span class="self">self</span>.<span class="ident">project_dir</span>, <span class="kw-2">&amp;</span><span class="self">self</span>.<span class="ident">project_parameters</span>.<span class="ident">source_dirs</span>)<span class="question-mark">?</span>;
        }

        <span class="kw">if</span> <span class="self">self</span>.<span class="ident">project_dir</span>.<span class="ident">is_file</span>() {
            <span class="kw">if</span> <span class="kw">let</span> <span class="prelude-val">Some</span>(<span class="ident">ext</span>) <span class="op">=</span> <span class="self">self</span>.<span class="ident">project_dir</span>.<span class="ident">extension</span>() {
                <span class="comment">// Project path is a file and has the right extension.</span>
                <span class="kw">if</span> <span class="ident">T</span>::<span class="ident">file_types</span>().<span class="ident">contains</span>(
                    <span class="kw-2">&amp;</span><span class="ident">ext</span>.<span class="ident">to_str</span>()
                        .<span class="ident">ok_or_else</span>(<span class="op">||</span> <span class="ident">err_msg</span>(<span class="string">&quot;Unable to stringify file extension.&quot;</span>))<span class="question-mark">?</span>,
                ) {
                    <span class="comment">// Push it to the project file vector for analyzing purposes.</span>
                    <span class="self">self</span>.<span class="ident">analysis</span>
                        .<span class="ident">project_files_mut</span>()
                        .<span class="ident">push</span>(<span class="ident">ProjectFile</span>::<span class="ident">new</span>(<span class="kw-2">&amp;</span><span class="self">self</span>.<span class="ident">project_dir</span>));
                }
            }
        }

        <span class="self">self</span>.<span class="ident">analysis</span>.<span class="ident">extract_entities</span>()<span class="question-mark">?</span>;

        <span class="prelude-val">Ok</span>(())
    }

    <span class="doccomment">/// Starts the synthesis of the target projects test files.</span>
    <span class="kw">pub</span> <span class="kw">fn</span> <span class="ident">synthesize</span><span class="op">&lt;</span><span class="ident">P</span>: <span class="ident">Into</span><span class="op">&lt;</span><span class="ident">PathBuf</span><span class="op">&gt;&gt;</span>(<span class="kw-2">&amp;</span><span class="kw-2">mut</span> <span class="self">self</span>, <span class="ident">base_path</span>: <span class="ident">P</span>) <span class="op">-&gt;</span> <span class="ident">Fallible</span><span class="op">&lt;</span>()<span class="op">&gt;</span> {
        <span class="self">self</span>.<span class="ident">synthesis</span>.<span class="ident">parse_stubs</span>(
            <span class="ident">DEFAULT_ENV_YML</span>,
            <span class="self">self</span>.<span class="ident">project_parameters</span>.<span class="ident">test_env</span>.<span class="ident">as_str</span>(),
            <span class="kw-2">&amp;</span><span class="ident">base_path</span>.<span class="ident">into</span>(),
        )<span class="question-mark">?</span>;

        <span class="kw">for</span> <span class="ident">file</span> <span class="kw">in</span> <span class="self">self</span>.<span class="ident">analysis</span>.<span class="ident">project_files</span>().<span class="ident">iter</span>() {
            <span class="self">self</span>.<span class="ident">synthesis</span>.<span class="ident">process_testfile</span>(<span class="ident">file</span>)<span class="question-mark">?</span>;
        }

        <span class="prelude-val">Ok</span>(())
    }
}
</pre>
</section><section id="search" class="content hidden"></section><section class="footer"></section><aside id="help" class="hidden"><div><h1 class="hidden">Help</h1><div class="shortcuts"><h2>Keyboard Shortcuts</h2><dl><dt><kbd>?</kbd></dt><dd>Show this help dialog</dd><dt><kbd>S</kbd></dt><dd>Focus the search field</dd><dt><kbd>↑</kbd></dt><dd>Move up in search results</dd><dt><kbd>↓</kbd></dt><dd>Move down in search results</dd><dt><kbd>↹</kbd></dt><dd>Switch tab</dd><dt><kbd>&#9166;</kbd></dt><dd>Go to active search result</dd><dt><kbd>+</kbd></dt><dd>Expand all sections</dd><dt><kbd>-</kbd></dt><dd>Collapse all sections</dd></dl></div><div class="infos"><h2>Search Tricks</h2><p>Prefix searches with a type followed by a colon (e.g. <code>fn:</code>) to restrict the search to a given type.</p><p>Accepted types are: <code>fn</code>, <code>mod</code>, <code>struct</code>, <code>enum</code>, <code>trait</code>, <code>type</code>, <code>macro</code>, and <code>const</code>.</p><p>Search functions by type signature (e.g. <code>vec -> usize</code> or <code>* -> vec</code>)</p><p>Search multiple things at once by splitting your query with comma (e.g. <code>str,u8</code> or <code>String,struct:Vec,test</code>)</p></div></div></aside><script>window.rootPath = "../../";window.currentCrate = "thinlinelib";</script><script src="../../aliases.js"></script><script src="../../main.js"></script><script defer src="../../search-index.js"></script></body></html>