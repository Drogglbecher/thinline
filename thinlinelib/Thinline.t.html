<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=struct.Thinline.html">
</head>
<body>
    <p>Redirecting to <a href="struct.Thinline.html">struct.Thinline.html</a>...</p>
    <script>location.replace("struct.Thinline.html" + location.search + location.hash);</script>
</body>
</html>