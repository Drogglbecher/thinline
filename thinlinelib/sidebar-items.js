initSidebarItems({"mod":[["analysis",""],["entity",""],["language_type",""],["project_parameters",""],["stubs",""],["synthesis",""],["value_parser",""]],"struct":[["Thinline","Global structure representing the `Thinline` lib."]]});