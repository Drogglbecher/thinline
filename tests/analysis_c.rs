extern crate snapshot;
extern crate thinlinelib;

#[cfg(test)]
mod c {
    use snapshot::snapshot;
    use std::path::Path;
    use thinlinelib::{analysis::Analysis, entity::EntityType, language_type::C};

    #[test]
    fn collect_sources() {
        // should_succeed
        {
            // when_directory_is_valid()
            {
                // Given
                let analysis: Analysis<C> = Analysis::new();

                // When
                let c_test_src_path = Path::new("tests").join("testdata").join("c_sources");
                assert!(
                    analysis
                        .collect_sources(&c_test_src_path, &[String::from(".")])
                        .is_ok()
                );

                // Then
                assert_eq!(analysis.project_files().len(), 5);
            }
        }

        // should_fail
        {
            // when_directory_not_existing()
            {
                // Given
                let analysis: Analysis<C> = Analysis::new();

                // When
                let c_test_src_path = Path::new("not").join("existing");

                // Then
                assert!(
                    analysis
                        .collect_sources(&c_test_src_path, &[String::from(".")])
                        .is_err()
                );
            }

            // when_path_is_no_directory()
            {
                // Given
                let analysis: Analysis<C> = Analysis::new();

                // When
                let c_test_src_path = Path::new("tests").join("lib.rs");

                // Then
                assert!(
                    analysis
                        .collect_sources(&c_test_src_path, &[String::from(".")])
                        .is_err()
                );
            }
        }
    }

    fn extract_entities_c() -> Vec<EntityType> {
        let analysis: Analysis<C> = Analysis::new();
        let c_test_src_path = Path::new("tests").join("testdata").join("analysis");
        assert!(
            analysis
                .collect_sources(&c_test_src_path, &[String::from(".")])
                .is_ok()
        );

        assert!(analysis.extract_entities().is_ok());

        let project_files = analysis.project_files();

        assert_eq!(project_files.len(), 1);

        let entities = project_files[0].entities();
        assert_eq!(entities.len(), 1);
        entities[0].clone().entities
    }

    #[cfg(target_os = "linux")]
    #[snapshot]
    fn extract_entities_linux_c() -> Vec<EntityType> {
        extract_entities_c()
    }

    #[cfg(target_os = "windows")]
    #[snapshot]
    fn extract_entities_windows_c() -> Vec<EntityType> {
        extract_entities_c()
    }
}
