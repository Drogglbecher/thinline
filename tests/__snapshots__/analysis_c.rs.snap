{
  "analysis_c::c::extract_entities_linux_c": {
    "file": "tests/analysis_c.rs",
    "module_path": "analysis_c::c",
    "test_function": "extract_entities_linux_c",
    "recorded_value": [
      {
        "Function": {
          "arguments": [
            {
              "atype": "int",
              "name": "no1",
              "value": null
            },
            {
              "atype": "int",
              "name": "no2",
              "value": null
            }
          ],
          "description": {
            "lines": [
              "#TESTCASE(Source1::CheckIfSumWorks)",
              "int test_no = 2;",
              "#EQ[TL_FCT(no1:test_no,no2:5)=>7]",
              "#EQ[TL_FCT(no1:5,no2:2)=>7]",
              "EXPECT_EQ(11, test_int_no1(9, 2));"
            ]
          },
          "name": "test_int_no1",
          "return_type": "int"
        }
      },
      {
        "Function": {
          "arguments": [
            {
              "atype": "const int *const",
              "name": "no1",
              "value": null
            },
            {
              "atype": "const int *const",
              "name": "no2",
              "value": null
            }
          ],
          "description": {
            "lines": [
              "#TESTCASE(Source1::TestPtr)",
              "int test_no = 2;",
              "int test_no2 = 5;",
              "#EQ[TL_FCT(no1:&test_no,no2:&test_no2)=>7]"
            ]
          },
          "name": "test_ptr",
          "return_type": "int"
        }
      },
      {
        "Function": {
          "arguments": [],
          "description": {
            "lines": [
              "#TESTCASE(Source1::EmptyFct)",
              "#EQ[TL_FCT()=>7]",
              "#NE[TL_FCT()=>4]"
            ]
          },
          "name": "test_empty_fct",
          "return_type": "int"
        }
      },
      {
        "Function": {
          "arguments": [
            {
              "atype": "const int",
              "name": "argc",
              "value": null
            },
            {
              "atype": "char *const []",
              "name": "argv",
              "value": null
            }
          ],
          "description": {
            "lines": [
              "This function has parameters, yeah"
            ]
          },
          "name": "main",
          "return_type": "int"
        }
      }
    ]
  },
  "analysis_c::c::extract_entities_windows_c": {
    "file": "tests\\analysis_c.rs",
    "module_path": "analysis_c::c",
    "test_function": "extract_entities_windows_c",
    "recorded_value": [
      {
        "Function": {
          "arguments": [
            {
              "atype": "int",
              "name": "no1",
              "value": null
            },
            {
              "atype": "int",
              "name": "no2",
              "value": null
            }
          ],
          "description": {
            "lines": [
              "#TESTCASE(Source1::CheckIfSumWorks)",
              "int test_no = 2;",
              "#EQ[TL_FCT(no1:test_no,no2:5)=>7]",
              "#EQ[TL_FCT(no1:5,no2:2)=>7]",
              "EXPECT_EQ(11, test_int_no1(9, 2));"
            ]
          },
          "name": "test_int_no1",
          "return_type": "int"
        }
      },
      {
        "Function": {
          "arguments": [
            {
              "atype": "const int *const",
              "name": "no1",
              "value": null
            },
            {
              "atype": "const int *const",
              "name": "no2",
              "value": null
            }
          ],
          "description": {
            "lines": [
              "#TESTCASE(Source1::TestPtr)",
              "int test_no = 2;",
              "int test_no2 = 5;",
              "#EQ[TL_FCT(no1:&test_no,no2:&test_no2)=>7]"
            ]
          },
          "name": "test_ptr",
          "return_type": "int"
        }
      },
      {
        "Function": {
          "arguments": [],
          "description": {
            "lines": [
              "#TESTCASE(Source1::EmptyFct)",
              "#EQ[TL_FCT()=>7]",
              "#NE[TL_FCT()=>4]"
            ]
          },
          "name": "test_empty_fct",
          "return_type": "int"
        }
      },
      {
        "Function": {
          "arguments": [
            {
              "atype": "const int",
              "name": "argc",
              "value": null
            },
            {
              "atype": "char *const []",
              "name": "argv",
              "value": null
            }
          ],
          "description": {
            "lines": [
              "This function has parameters, yeah"
            ]
          },
          "name": "main",
          "return_type": "int"
        }
      }
    ]
  }
}
